src/modules/sensors/vehicle_imu/VehicleIMU.cpp

```diff
@@ -85,52 +93,259 @@ void VehicleIMU::ParametersUpdate(bool force)
 
-		// publich vehicle_imu
-		vehicle_imu_s imu;
-		imu.timestamp_sample = accel.timestamp_sample;
-		imu.accel_device_id = accel.device_id;
-		imu.gyro_device_id = gyro.device_id;
+		if (_accel_integrator.reset(delta_velocity, accel_integral_dt)
+		    && _gyro_integrator.reset(delta_angle, gyro_integral_dt)) {
 
-		delta_angle.copyTo(imu.delta_angle);
-		delta_velocity.copyTo(imu.delta_velocity);
+			// publich vehicle_imu
+			vehicle_imu_s imu{};
 
-		imu.dt = accel.dt;
-		//imu.clip_count = accel.clip_count;
-		imu.timestamp = hrt_absolute_time();
+			imu.timestamp_sample = (_last_timestamp_sample_accel + _last_timestamp_sample_gyro) / 2;
+			imu.accel_device_id = _accel_corrections.get_device_id();
+			imu.gyro_device_id = _gyro_corrections.get_device_id();
+			delta_angle.copyTo(imu.delta_angle);
+			delta_velocity.copyTo(imu.delta_velocity);
+			imu.delta_angle_dt = gyro_integral_dt;
+			imu.delta_velocity_dt = accel_integral_dt;
+			imu.delta_velocity_clipping = _delta_velocity_clipping;
+			imu.timestamp = hrt_absolute_time();
 
-		_vehicle_imu_pub.publish(imu);
+			_vehicle_imu_pub.publish(imu);
+
+			perf_count_interval(_publish_interval_perf, imu.timestamp);
+
+			UpdateAccelVibrationMetrics(delta_velocity);
+			UpdateGyroVibrationMetrics(delta_angle);
+
+			PublishStatus();
+
+			// reset
+			_delta_velocity_clipping = 0;
+		}
 	}
 }
 
+void VehicleIMU::UpdateIntergratorConfiguration()
+{
+	const float configured_interval_us = 1e6f / _param_imu_integ_rate.get();
+
+	if ((_accel_interval.update_interval > 0) && (_gyro_interval.update_interval > 0)) {
+
+		// determine closest number of sensor samples that will get closest to the desired integration interval
+		const uint8_t integral_samples = math::constrain((uint8_t)roundf(configured_interval_us /
+						 _gyro_interval.update_interval), (uint8_t)1, sensor_gyro_s::ORB_QUEUE_LENGTH);
+
+		const float interval_us = roundf(integral_samples * _gyro_interval.update_interval);
+
+		// relaxed minimum reset interval (closest we can get to configured interval minus half a sample interval to tolerate any jitter)
+		const float relaxed_reset_interval_min = roundf(interval_us - 0.5f * _gyro_interval.update_interval);
+
+		_gyro_integrator.set_reset_samples(integral_samples);
+		_gyro_integrator.set_reset_interval(relaxed_reset_interval_min);
+
+		_sensor_accel_sub.set_required_updates(integral_samples);
+		_sensor_gyro_sub.set_required_updates(integral_samples);
+
+		if ((_accel_corrections.get_device_id() == _gyro_corrections.get_device_id())
+		    && fabsf(_accel_interval.update_interval - _gyro_interval.update_interval) < 10) {
+
+			// if accel and gyro are coming form the same device and data rates are extremely close we can schedule as an integer multipe of samples
+			_accel_integrator.set_reset_samples(integral_samples);
+			_accel_integrator.set_reset_interval(relaxed_reset_interval_min);
+
+		} else {
+			// otherwise let the gyro set the configuration and scheduling
```

Comment:
This started life implemented for the known good/simple cases (already synchronized invensense accel + gyro) and then evolved to accommodate the others. I think always following the gyro makes sense and keeps the scheduling regular relative to the inner loop.

Reply:
Lockstep and SITL (non-root) was being a little weird with the possibility of gyro then accel publishing, but VehicleIMU running in between. I'll see if it can be simplified (sensor_gyro only scheduling).